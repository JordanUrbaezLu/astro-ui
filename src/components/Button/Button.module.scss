@use "../../utility/globals";

.button {
  align-items: center;
  border: 2px solid globals.$primary-background-color;
  cursor: pointer;
  display: flex;
  outline: none;
}

.primary {
  background-color: globals.$primary-background-color;

  &:hover,
  &:focus {
    background-color: globals.$primary-background-color-hover-and-focus;
    border: 2px solid
      globals.$primary-background-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$primary-background-color-active;
    border: 2px solid globals.$primary-background-color-active;
  }
}

.secondary {
  background-color: white;
  border: 2px solid globals.$secondary-text-color;

  &:hover,
  &:focus {
    background-color: globals.$secondary-background-color-hover-and-active;
    border: 2px solid globals.$secondary-border-color-hover-and-focus;
    color: globals.$secondary-text-color-hover-and-focus;
  }

  &:active {
    background-color: globals.$secondary-background-color-active;
  }
}

.small {
  border-radius: 6px;
  height: 25.8px;
  padding: 4px 6px;

  .leadingIcon,
  .trailingIcon {
    height: 12px;
  }
}

.medium {
  border-radius: 8px;
  height: 34.5px;
  padding: 6px 9px;

  .leadingIcon,
  .trailingIcon {
    height: 16px;
  }
}

.large {
  border-radius: 10px;
  height: 43px;
  padding: 8px 12px;

  .leadingIcon,
  .trailingIcon {
    height: 20px;
  }
}

.smallRipple {
  border-radius: 6px;
}

.mediumRipple {
  border-radius: 8px;
}

.largeRipple {
  border-radius: 10px;
}

.typography {
  align-items: center;
  display: flex;
  gap: 6px;
}

.disableRipple {
  pointer-events: none !important;
}

.disabled {
  background-color: lightgray !important;
  border: 2px solid lightgray;
  cursor: default !important;
  user-select: none !important;
}
