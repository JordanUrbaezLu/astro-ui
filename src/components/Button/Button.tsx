import * as React from "react";
import styles from "./Button.module.scss";
import classNames from "classnames";
import { Typography } from "../Typography/Typography";
import { Ripple } from "../Ripple";

export type ButtonVariant = "primary" | "secondary";
export type ButtonSize = "small" | "medium" | "large";
export type ButtonType = "button" | "submit" | "reset";

export interface ButtonProps
  extends React.ComponentPropsWithoutRef<"button"> {
  /**
   * The content for the Button
   */
  children: string;
  /**
   * If the Button is disabled
   *
   * @default false
   */
  disabled?: boolean;
  /**
   * The leading icon for the Button
   */
  leadingIcon?: React.ReactNode;
  /**
   * The callback fired when the Button is clicked
   */
  onClick?: () => void;
  /**
   * The size of the Button
   *
   * @default "medium"
   */
  size?: ButtonSize;
  /**
   * The trailing icon for the Button
   */
  trailingIcon?: React.ReactNode;
  /**
   * The type of the Button
   */
  type?: ButtonType;
  /**
   * The variant of the Button
   *
   * @default "primary"
   */
  variant?: ButtonVariant;
}

/**
 * @public
 *
 * @description
 *
 * The Button component is used to allows user to take an action.
 */
export const Button = React.forwardRef<
  HTMLButtonElement,
  ButtonProps
>((props, ref) => {
  const {
    className,
    children,
    disabled = false,
    leadingIcon,
    onClick,
    size = "medium",
    trailingIcon,
    type,
    variant = "primary",
    ...rest
  } = props;

  return (
    <button
      className={classNames(
        className,
        styles.button,
        disabled && styles.disabled,
        size === "small"
          ? styles.small
          : size === "large"
          ? styles.large
          : styles.medium,
        variant === "secondary" ? styles.secondary : styles.primary
      )}
      disabled={disabled}
      onClick={onClick}
      ref={ref}
      type={
        type === "reset"
          ? "reset"
          : type === "submit"
          ? "submit"
          : "button"
      }
      {...rest}
    >
      {!disabled && <Ripple position="cursor" />}
      <Typography
        className={styles.typography}
        type={
          size === "large" ? "h1" : size === "medium" ? "h2" : "h3"
        }
        variant={variant}
      >
        {leadingIcon && (
          <span className={styles.leadingIcon}>{leadingIcon}</span>
        )}
        {children}
        {trailingIcon && (
          <span className={styles.trailingIcon}>{trailingIcon}</span>
        )}
      </Typography>
    </button>
  );
});

Button.displayName = "Button";
