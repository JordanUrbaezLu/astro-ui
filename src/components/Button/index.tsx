/* istanbul ignore file */
export * from "./Button";
