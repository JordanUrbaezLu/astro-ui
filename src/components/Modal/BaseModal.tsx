import * as React from "react";
import styles from "./BaseModal.module.scss";
import { IconButton } from "../IconButton/IconButton";
import { useOnClickOutside, useOnKeyDown } from "../../hooks";
import { TbX } from "react-icons/tb";
import { FocusTrap } from "../FocusTrap/FocusTrap";
import { Divider } from "../Divider/Divider";
import classNames from "classnames";

export interface BaseModalProps
  extends React.ComponentPropsWithoutRef<"div"> {
  /**
   * The accessible label for the Modal
   */
  ariaLabel?: string;
  /**
   * The buttons for the Modal
   */
  buttons?: React.ReactNode;
  /**
   * The content for the Modal
   */
  content: string;
  /**
   * The callback fired when the Modal closes
   */
  onClose: () => void;
  /**
   * The title for the Modal
   */
  title: string;
}

/**
 * @private
 */
export const BaseModal: React.FC<BaseModalProps> = (props) => {
  const {
    ariaLabel,
    buttons,
    className,
    content,
    onClose,
    title,
    ...rest
  } = props;

  const modalRef = React.useRef<HTMLDivElement>(null);

  useOnClickOutside(modalRef, onClose);

  useOnKeyDown(["Escape"], onClose);

  return (
    <FocusTrap>
      <div
        aria-label={ariaLabel}
        className={classNames(className, styles.modalContainer)}
        ref={modalRef}
        role="dialog"
        {...rest}
      >
        <div className={styles.modalHeader}>
          <div className={styles.modalTitle}>{title}</div>
          <IconButton
            className={styles.modalIcon}
            content={<TbX />}
            onClick={onClose}
          />
        </div>
        <div className={styles.modalContent}>{content}</div>
        {buttons && (
          <>
            <Divider direction="horizontal" stroke={1} />
            <div className={styles.modalButtons}>{buttons}</div>
          </>
        )}
      </div>
    </FocusTrap>
  );
};
