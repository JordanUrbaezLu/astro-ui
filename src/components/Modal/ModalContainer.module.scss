@use "../../utility/globals";

.modalContainer {
  background: globals.$primary-background-color;
  border-radius: 10px;
  box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
  width: 400px;
  color: globals.$primary-text-color;
}

.modalHeader {
  box-sizing: border-box;
  display: flex;
  padding: 15px 15px 0px 15px;
  justify-content: space-between;
  width: 100%;
}

.modalContent {
  padding: 30px 30px 50px 30px;
  text-align: center;
}

.modalButtons {
  display: flex;
  width: 100%;
  justify-content: end;
  padding: 15px;
  box-sizing: border-box;
}

.modalTitle {
  align-self: center;
}

.modalIcon {
  color: globals.$primary-text-color !important;

  &:hover,
  &:focus {
    background-color: globals.$primary-background-color-hover-and-focus !important;
  }

  &:active {
    background-color: globals.$primary-background-color-active !important;
  }
}
