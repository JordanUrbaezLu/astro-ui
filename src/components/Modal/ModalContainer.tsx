import * as React from "react";
import styles from "./ModalContainer.module.scss";
import { IconButton } from "../IconButton/IconButton";
import { useOnClickOutside } from "../../hooks";
import { TbX } from "react-icons/tb";
import { useLockBodyScroll } from "../../hooks/useLockBodyScroll";
import FocusTrap from "focus-trap-react";

export interface ModalContainerProps {
  /**
   * The buttons for the Modal
   */
  buttons?: React.ReactNode;
  /**
   * The content for the Modal
   */
  content: string;
  /**
   * If the Modal is open
   */
  isOpen: boolean;
  /**
   * The callback fired when the Modal closes
   */
  onClose: () => void;
  /**
   * The title for the Modal
   */
  title: string;
}

export const ModalContainer: React.FC<ModalContainerProps> = (props) => {
  const { buttons, content, isOpen, onClose, title } = props;

  const modalRef = React.useRef<HTMLDivElement>(null);

  useLockBodyScroll(isOpen);

  useOnClickOutside(modalRef, onClose);

  return (
    <FocusTrap>
      <div className={styles.modalContainer} ref={modalRef}>
        <div className={styles.modalHeader}>
          <div className={styles.modalTitle}>{title}</div>
          <IconButton content={<TbX />} onClick={onClose}/>
        </div>
        <div className={styles.modalContent}>{content}</div>
        {buttons && <div className={styles.modalButtons}>{buttons}</div>}
      </div>
    </FocusTrap>
  );
};
