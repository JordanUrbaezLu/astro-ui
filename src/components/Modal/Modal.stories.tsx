import * as React from "react";
import { Modal } from "./Modal";
import { Button } from "../Button/Button";
import { ButtonSet } from "../ButtonSet";

export default {
  component: Modal,
  title: "Components/Modal",
};

export const Default = () => {
  const [isOpen, setIsOpen] = React.useState<boolean>(true);

  return (
    <Modal
      content="Modal Content"
      isOpen={isOpen}
      onOpen={() => setIsOpen(true)}
      onClose={() => setIsOpen(false)}
      trigger={<Button>Open Modal!</Button>}
      title="Modal Title"
    />
  );
};

export const WithButtonSet = () => {
  const [isOpen, setIsOpen] = React.useState<boolean>(true);

  return (
    <Modal
      buttons={
        <ButtonSet>
          <Button>Cancel</Button>
          <Button>Confirm</Button>
        </ButtonSet>
      }
      content="Modal Content"
      isOpen={isOpen}
      onOpen={() => setIsOpen(true)}
      onClose={() => setIsOpen(false)}
      trigger={<Button>Open Modal!</Button>}
      title="Modal Title"
    />
  );
};

export const WithNestedModals = () => {
  const [isFirstModalOpen, setIsFirstModalOpen] =
    React.useState<boolean>(true);
  const [isSecondModalOpen, setSecondModalIsOpen] =
    React.useState<boolean>(true);

  return (
    <Modal
      buttons={
        <ButtonSet>
          <Button>Cancel</Button>
          <Modal
            buttons={
              <ButtonSet>
                <Button>Cancel</Button>
                <Button>Confirm</Button>
              </ButtonSet>
            }
            content="Modal Content"
            hasOverlay={false}
            isOpen={isSecondModalOpen}
            onOpen={() => setSecondModalIsOpen(true)}
            onClose={() => setSecondModalIsOpen(false)}
            trigger={<Button>Open Second Modal!</Button>}
            title="Modal Title"
          />
        </ButtonSet>
      }
      content="Modal Content"
      isOpen={isFirstModalOpen}
      onOpen={() => setIsFirstModalOpen(true)}
      onClose={() => setIsFirstModalOpen(false)}
      trigger={<Button>Open First Modal!</Button>}
      title="Modal Title"
    />
  );
};
