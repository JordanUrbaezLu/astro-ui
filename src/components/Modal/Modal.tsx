import * as React from "react";
import styles from "./Modal.module.scss";
import { Overlay } from "../Overlay/Overlay";
import { ModalContainer } from "./ModalContainer";
import { CSSTransition } from "react-transition-group";

export interface ModalProps {
  /**
   * The buttons for the Modal
   */
  buttons?: React.ReactNode;
  /**
   * The content for the Modal
   */
  content: string;
  /**
   * If the Modal has an overlay
   *
   * @default true
   */
  hasOverlay?: boolean;
  /**
   * If the Modal is open
   *
   * @default false
   */
  isOpen: boolean;
  /**
   * The callback fired when the Modal closes
   */
  onClose: () => void;
  /**
   * The callback fired when the Modal opens
   */
  onOpen: () => void;
  /**
   * The title for the Modal
   */
  title: string;
  /**
   * The trigger for the Modal
   */
  trigger: React.ReactElement;
}

export const Modal: React.FC<ModalProps> = (props) => {
  const {
    buttons,
    content,
    hasOverlay = true,
    isOpen = false,
    onClose,
    onOpen,
    title,
    trigger,
  } = props;

  return (
    <>
      {React.cloneElement(trigger, {
        onClick: onOpen,
        role: "button",
        tabIndex: 0,
      })}
      <CSSTransition
        in={isOpen}
        timeout={250}
        mountOnEnter
        unmountOnExit
        classNames={{
          enter: styles.enter,
          enterActive: styles.enterActive,
          exit: styles.exit,
          exitActive: styles.exitActive,
        }}
      >
        <Overlay showBackground={hasOverlay}>
          <div className={styles.modalBackground}>
            <ModalContainer
              buttons={buttons}
              content={content}
              isOpen={isOpen}
              onClose={onClose}
              title={title}
            />
          </div>
        </Overlay>
      </CSSTransition>
    </>
  );
};
