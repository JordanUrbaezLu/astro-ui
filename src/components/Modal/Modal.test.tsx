import { render, screen } from "@testing-library/react";
import { Modal } from "./Modal";
import { Button } from "../Button/Button";
import { axe } from "jest-axe";

describe("Modal", () => {
  test("Should render component correctly.", () => {
    const { container } = render(
      <Modal
        content="Modal Content"
        isOpen={false}
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
        title="Modal Title"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe("Accessibility", () => {
  test("Should have no accessibility violations.", async () => {
    const { container } = render(
      <Modal
        content="Modal Content"
        isOpen={false}
        onOpen={jest.fn}
        onClose={jest.fn()}
        trigger={<Button>Trigger</Button>}
        title="Modal Title"
      />
    );
    const results = axe(container, {
      rules: { tabindex: { enabled: false } },
    });
    expect(await results).toHaveNoViolations();
  });
});

// describe("Props", () => {
//   test("Should render content correctly.", () => {
//     render(
//       <Modal
//         content="Modal Content"
//         isOpen={false}
//         onOpen={jest.fn}
//         onClose={jest.fn()}
//         trigger={<Button>Trigger</Button>}
//         title="Modal Title"
//       />
//     );
//     expect(screen.getByText("Modal Content")).toBeInTheDocument();
//   });

//   test("Should render title correctly.", () => {
//     render(
//       <Modal
//         content="Modal Content"
//         isOpen={false}
//         onOpen={jest.fn}
//         onClose={jest.fn()}
//         trigger={<Button>Trigger</Button>}
//         title="Modal Title"
//       />
//     );
//     expect(screen.getByText("Modal Title")).toBeInTheDocument();
//   });
// });
