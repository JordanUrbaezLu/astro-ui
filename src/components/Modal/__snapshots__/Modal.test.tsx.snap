// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Should render component correctly. 1`] = `
<button
  class="button medium primary"
  role="button"
  tabindex="0"
  type="button"
>
  Trigger
</button>
`;
